### ISSUE TITLE
🐛 Bug: remove util.inherits / convert to classes

### ISSUE THREAD

**Issue by jimmywarting** [mochajs/mocha#5025]
**Is your feature request related to a problem or a nice-to-have?? Please describe.**
It would be nice to get rid of some NodeJS specific things that makes it better suited for browsers.
one such thing could be the option to remove `util.inherits` and replacing it with classes that can extends things.
resulting in fewer dependencies and smaller code bundle that won't require the browserified version of [util](https://npmgraph.js.org/?q=util#select=exact%3Autil%400.12.5) that has an additional of 20 modules and 12 kb

NodeJS docs have marked it as legacy and has this to say about `util.inherits`

> Usage of util.inherits() is discouraged. Please use the ES6 class and extends keywords to get language level inheritance support

**Describe the solution you'd like**
Change prototype based classes into es-classes

**Comment by voxpelli** [comment]
Fully agree!

Some classes that needs to be converted:

* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/hook.js#L29
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/runnable.js#L55
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/suite.js#L99
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/test.js#L39
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/dot.js#L79
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/landing.js#L114
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/list.js#L76
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/min.js#L50
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/nyan.js#L78
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/progress.js#L102
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/spec.js#L99
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/tap.js#L80
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/tap.js#L239
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/tap.js#L291
* https://github.com/mochajs/mocha/blob/24560c1532b8e67d1254b489fcc6f84c4033c0e1/lib/reporters/xunit.js#L117

If we can find a linting rule that enforces / helps with this, then we should

**Comment by JoshuaKGoldberg** [comment]
Bad news: I think this would technically be a breaking change. Out of the three PRs that split up:

* #5178: no changes needed for tests
* #5179: no changes needed for tests
* #5180: quite a few changes needed for tests

...indicating that the runtime differences noted in https://nodejs.org/api/util.html#utilinheritsconstructor-superconstructor do impact usage of these class(-like)s.

These are all part of the public Mocha API:

* https://mochajs.org/api/hook
* https://mochajs.org/api/suite
* https://mochajs.org/api/runnable
* https://mochajs.org/api/test
* https://mochajs.org/api/module-base

As much as we want to clean up this old code, we'll have to wait for the next major version of Mocha. 😞

**Comment by voxpelli** [comment]
I think we should get planning for a new major version fairly soon, as it eg is needed to update some outdated dependencies as well

We can keep a 10.x branch around for bug fixes and have main be for the next major?

**Comment by JoshuaKGoldberg** [comment]
> get planning for a new major version fairly soon

Agreed, the more I dig into the legacy dependencies the more breaking changes I find...

For visibility: we'll talk about this in our next maintainer meeting to see what strategy we go with. 👍

**Comment by samuelhnrq** [comment]
Is this accepting PRs? I'd like a shot at this one. It doesn't have the label 👀

**Comment by JoshuaKGoldberg** [comment]
Good question. It's kind of stuck in limbo right now: as a breaking change, we're going to have to wait for the next major (v12). That's blocked on https://github.com/mochajs/mocha/issues/5357#issuecomment-3035961293:

> I'd like to wait a bit longer for Mocha 12 to let the new Node.js support stabilize.

You could send a PR if you wanted to, but we wouldn't be able to merge it until the PRs for v12 start to land likely later this year. We can review it, though, if you'd like to get started sooner. 🙂 

Edit: oh, and see https://github.com/mochajs/mocha/issues/5025#issuecomment-2241154024 - you can probably reuse some existing work if you'd like.

**Comment by JoshuaKGoldberg** [comment]
@samuelhnrq we're now merging Mocha 12 PRs! Do you still want to tackle this one?

**Comment by Offroaders123** [comment]
Okay, I think it's all covered! Let me know if there are any I missed. I searched for `prototype` and `inherits`, and everything seems to be handled now.

Edit: Having some trouble with ES2022 syntax which I hadn't realized wasn't supported by the tooling we have. Will keep you updated!

**Comment by Offroaders123** [comment]
Discovering that it doesn't seem that Rollup is agreeing with the new changes. It appears to be with the `static fieldName = "this is a field"` syntax. What flavor of ES6 is it compatible with using?

**Comment by Offroaders123** [comment]
Okay, the previous PR was getting a bit convoluted. Going to start fresh to some extent to ensure everything is passing before trying to merge [self-aware noob skill issue 🤓👆]

I think the changes themselves aren't necessarily too convoluted, but I didn't properly test them properly to solve what wasn't working.

**Comment by JoshuaKGoldberg** [comment]
@Offroaders123 are you still planning on working on this?

**Comment by JoshuaKGoldberg** [comment]
This is a nice cleanup for v12 that will make future work such as the internal CJS->ESM migration (#5400) a little easier. I'll take it on.

**Comment by mark-wiemer** [comment]
Fixed up https://github.com/mochajs/mocha/pull/5178, going to see if we can have Copilot help out here

**Comment by mark-wiemer** [comment]
nvm, we already have https://github.com/mochajs/mocha/pull/5180 :)

**Comment by mark-wiemer** [comment]
@mochajs/maintenance-crew, just want to call out that this is indeed a breaking change and as such we should re-evaluate whether it should land in 12 ("no intentional breaking changes", #5357 ) or 13. It'd be a lot of work to undo, but we'd maintain consistency with our even-odd approach.

Personally, I think we should relax our even-odd approach at this point, we have had a few "accidental" breaking changes that we've worked to undo even though they're very minimal.

That said, we also clearly stated that we'd remove `util.inherits` in #5357 but didn't communicate that it would in fact be a breaking change for custom reporters.

Ref #5677 and related commits for details.

**Comment by mark-wiemer** [comment]
Going to re-open this one at least until we have clear docs :/

**Comment by voxpelli** [comment]
> Personally, I think we should relax our even-odd approach at this point, we have had a few "accidental" breaking changes that we've worked to undo even though they're very minimal.

I agree with this. I think it should be about intent and purpose. If the purpose is to introduce a breaking change or whether it’s a justifiable side effect of modernization.

(Kind of applies to #5835 as well)

**Comment by mark-wiemer** [comment]
(The other related PR is #5409)

**Comment by JoshuaKGoldberg** [comment]
Agreed, this might be a technical breaking change but that's a side effect. The true intent is modernization. IMO this should stay in v12.

**Comment by mark-wiemer** [comment]
I think we're aligned this will stay in v12, now we just need clear migration docs, both for this bug to be closed and for our announcement :)

**Comment by mark-wiemer** [comment]
Our explainers page links to https://github.com/mochajs/mocha-examples/tree/main/packages/third-party-reporter, we can use that to help us write a migration guide too :)
